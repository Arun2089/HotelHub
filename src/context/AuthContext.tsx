import { createContext, useState, PropsWithChildren, useEffect } from "react";
import axios from "axios";
import Loader from "@components/Loader";
import AlertComponent from "@components/Alert";
import useAlert from "@src/hooks/useAlert";
import { UserDetailsType } from "@ts/types";
import {
  fetchData,
  getLocalStorage,
  deleteLocalStorage,
  setLocalStorage,
  catchErrorMessage,
} from "@utils/index";
import {
  GOOGLE_LOGIN_URL,
  USER_LOGIN_URL,
  USER_DATA_URL,
  ERROR,
  SUCCESS,
} from "@constant/index";

type AuthContextType = {
  user: UserDetailsType;
  updateUserDetails: (updatedUserDetails: UserDetailsType) => void;
  getUser: () => Promise<void>;
  isLoading: boolean;
  handleLogout: (showAlert: boolean) => void;
  handleLogin: (email: string, password: string) => void;
  handleGoogleLogin: () => void;
};

export const AuthContext = createContext<AuthContextType>({
  user: {
    id: null,
    name: "",
    email: "",
    role: "",
  },
  updateUserDetails: () => {},
  getUser: () => new Promise<void>(() => {}),
  isLoading: true,
  handleLogout: () => {},
  handleLogin: () => {},
  handleGoogleLogin: () => {},
});

const AuthProvider = ({ children }: PropsWithChildren) => {
  const [user, setUser] = useState<UserDetailsType | null>(null);
  const [isLoading, setisLoading] = useState<boolean>(true);
  const [error, setError] = useState<string>("");
  const { alert, handleAlert } = useAlert();
  const { isOpen, message, severity } = alert;

  useEffect(() => {
    getUser();
  }, []);

  const getUser = () => {
    return new Promise<void>((resolve, reject) => {
      const token = getLocalStorage("jwtToken");
      if (!token) {
        setisLoading(false);
        resolve();
        return;
      }
      fetchData(USER_DATA_URL + "?populate=*")
        .then((response) => {
          const {
            id,
            username,
            email,
            role: { type },
            student,
          } = response;
          setUser({
            id,
            name: student ? student.studentName : username,
            email,
            role: type,
          });
          resolve();
        })
        .catch((error) => {
          setError("Something went wrong");
          handleAlert(true, error.message, ERROR);
          reject(error);
        })
        .finally(() => {
          setisLoading(false);
        });
    });
  };

  function updateUserDetails(updatedUserDetails: UserDetailsType) {
    setUser(updatedUserDetails);
  }

  const handleLogin = async (email: string, password: string) => {
    try {
      const formData = new FormData();
      formData.append("identifier", email);
      formData.append("password", password);
      const response = await axios.post(USER_LOGIN_URL, formData, {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      });
      if (response.status === 200) {
        setLocalStorage("jwtToken", response.data.jwt);
        await getUser();
        handleAlert(true, "Logged in successfully", SUCCESS);
      } else {
        handleAlert(true, "Unable to Log in", ERROR);
      }
    } catch (error) {
      handleAlert(true, catchErrorMessage(error), ERROR);
    }
  };

  const handleGoogleLogin = () => {
    window.location.href = GOOGLE_LOGIN_URL;
  };

  function handleLogout(showAlert: boolean = true) {
    setUser(null);
    deleteLocalStorage("jwtToken");
    showAlert && handleAlert(true, "Logged out successfully", SUCCESS);
  }

  if (isLoading && !error) return <Loader />;

  return (
    <AuthContext.Provider
      value={{
        user: user!,
        updateUserDetails,
        isLoading,
        getUser,
        handleLogout,
        handleLogin,
        handleGoogleLogin,
      }}
    >
      {children}
      {isOpen && (
        <AlertComponent
          message={message}
          severity={severity}
          handleClose={() => handleAlert(false)}
        />
      )}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
