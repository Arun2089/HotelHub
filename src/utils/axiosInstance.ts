import axios, { AxiosInstance } from "axios";
import { BASE_URL } from "@constant/index";
import { deleteLocalStorage, getLocalStorage } from "@utils/index";

const axiosInstance = (): AxiosInstance => {
  const instance = axios.create({
    baseURL: BASE_URL,
  });
  instance.interceptors.request.use(
    (config) => {
      const token = getLocalStorage("jwtToken");
      if (token) config.headers.Authorization = `Bearer ${token}`;
      else window.location.href = "/login";
      return config;
    },
    (error) => Promise.reject(error)
  );

  instance.interceptors.response.use(
    (response) => response,
    (error) => {
      if (error.response.status === 401) {
        deleteLocalStorage("jwtToken");
        window.location.href = "/login";
      }
      return Promise.reject(error);
    }
  );

  return instance;
};

export default axiosInstance;
