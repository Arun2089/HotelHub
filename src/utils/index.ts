import { StudentInfoType } from "@ts/types";
import { STUDENT_INFO_URL } from "@constant/index";

export const fetchData = async (url: string) => {
  const response = await fetch(url);
  if (response.ok) {
    const data = await response.json();
    return data;
  } else {
    throw new Error(`Error in retrieving data from ${url}`);
  }
};
export const sendData = async (
  url: string,
  method: "POST" | "PUT",
  data: any //eslint-disable-line
): Promise<boolean> => {
  const response = await fetch(url, {
    method: method,
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });
  if (response.ok) {
    return true;
  } else {
    throw new Error(`Error in sending data`);
  }
};

export const deleteData = async (url: string): Promise<boolean> => {
  const response = await fetch(url, {
    method: "DELETE",
  });
  if (response.ok) {
    return true;
  } else {
    throw new Error(`Error in deleting data from ${url}`);
  }
};

export const checkEmailExists = async (
  email: string,
  id?: number
): Promise<boolean> => {
  try {
    const data = await fetchData(STUDENT_INFO_URL);
    if (id) {
      const emailExists = data.some(
        (student: StudentInfoType) =>
          student.email === email && student.id !== id
      );
      return emailExists;
    }
    const emailExists = data.some(
      (student: StudentInfoType) => student.email === email
    );
    return emailExists;
  } catch (error) {
    throw error as string;
  }
};

export const catchErrorMessage = (error: unknown) => {
  return error instanceof Error ? error.message : "An error occurred";
};
