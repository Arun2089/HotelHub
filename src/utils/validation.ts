function validateEmail(email: string) {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (email.trim() === "" || !emailRegex.test(email)) {
    return false;
  }
  return true;
}

function validatePhone(phone: string) {
  const phoneRegex = /^[6-9]\d{9}$/;
  if (phone.trim() === "" || !phoneRegex.test(phone)) {
    return false;
  }
  return true;
}

function validatePassword(password: string) {
  if (password.length < 8 || password.trim() === "") {
    return false;
  }
  return true;
}

function validateText(text: string) {
  const nameRegex = /^[A-Za-z\s]+$/;
  if (text.trim() === "" || !nameRegex.test(text)) {
    return false;
  }
  return true;
}

const validateError = (value: string, type: string) => {
  if (value === "") return false;
  switch (type) {
    case "text":
      return !validateText(value);
    case "email":
      return !validateEmail(value);
    case "tel":
      return !validatePhone(value);
    case "password":
    case "confirmPassword":
      return !validatePassword(value);
    default:
      return false;
  }
};
const getHelperText = (
  value: string,
  type: string,
  confirmPassword?: string
): string => {
  if (validateError(value, type)) {
    switch (type) {
      case "text":
        return "Name cannot have special characters";
      case "email":
        return "Enter a valid email";
      case "tel":
        return "Enter a valid phone number";
      case "password":
      case "confirmPassword": {
        return "Password must contain atleast 8 characters";
      }
      default:
        return "Invalid input";
    }
  } else if (type === "confirmPassword" && value && value !== confirmPassword) {
    return "Password does not match";
  } else return "";
};

export {
  validateEmail,
  validatePhone,
  validatePassword,
  validateText,
  validateError,
  getHelperText,
};
