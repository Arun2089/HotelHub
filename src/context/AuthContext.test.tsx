import { render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { AuthContext } from "./AuthContext";
import { fetchData } from "@src/utils";

jest.mock("@utils/index", () => ({
  fetchData: jest.fn().mockResolvedValue({
    user: {
      id: 1,
      username: "john@example.com",
      email: "john@example.com",
      role: { type: "admin" },
    },
  }),
  getLocalStorage: jest.fn().mockReturnValue("jwtToken"),
}));

describe("AuthContext", () => {
  test("SnapShot Test", () => {
    const { asFragment } = render(
      <AuthContext.Provider
        value={{
          user: {
            id: 1,
            name: "John Doe",
            email: "john@example.com",
            role: "admin",
          },
          updateUserDetails: jest.fn(),
          getUser: jest.fn(),
          isLoading: false,
          handleLogout: jest.fn(),
          handleLogin: jest.fn(),
          handleGoogleLogin: jest.fn(),
        }}
      >
        <div>Test Component</div>
      </AuthContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render the AuthProvider component", () => {
    render(
      <AuthContext.Provider
        value={{
          user: {
            id: 1,
            name: "John Doe",
            email: "john@example.com",
            role: "admin",
          },
          updateUserDetails: jest.fn(),
          getUser: jest.fn(),
          isLoading: false,
          handleLogout: jest.fn(),
          handleLogin: jest.fn(),
          handleGoogleLogin: jest.fn(),
        }}
      >
        <div>Test Component</div>
      </AuthContext.Provider>
    );

    const testComponent = screen.getByText("Test Component");
    expect(testComponent).toBeInTheDocument();
  });

  test("should call getUser function on mount", async () => {
    const getUserMock = jest.fn();

    render(
      <AuthContext.Provider
        value={{
          user: {
            id: 1,
            name: "John Doe",
            email: "john@example.com",
            role: "admin",
          },
          updateUserDetails: jest.fn(),
          getUser: getUserMock,
          isLoading: false,
          handleLogout: jest.fn(),
          handleLogin: jest.fn(),
          handleGoogleLogin: jest.fn(),
        }}
      >
        <div>Test Component</div>
      </AuthContext.Provider>
    );

    await waitFor(() => {
      expect(fetchData).toHaveBeenCalled();
    });
  });
});
