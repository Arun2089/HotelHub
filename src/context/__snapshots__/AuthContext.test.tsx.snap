// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthContext SnapShot Test 1`] = `
<DocumentFragment>
  <div>
    Test Component
  </div>
</DocumentFragment>
`;
